### PR TITLE
feature:  remove metrics when stopping task 

### LINKED ISSUE #503: UCP: remove metrics when stopping the task
## Feature Request

**Is your feature request related to a problem? Please describe:**

After stopped some tasks, their metrics still kept in Prometheus, this makes monitoring seem verbose. 

**Describe the feature you'd like:**

Remove metrics relevant to the task when stopping the task, like calling `metricVec.DeleteLabelValues`.

## Score 

500

## Mentor(s)
* @csuzhangxc 

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* Golang
* Experience with DM
* Experience with Prometheus

## Learning Materials

- [DM source reading](https://pingcap.com/blog-cn/#DM-%E6%BA%90%E7%A0%81%E9%98%85%E8%AF%BB)
- [DM document](https://pingcap.com/docs-cn/stable/reference/tools/data-migration/overview/#dm-%e6%9e%b6%e6%9e%84)

### PR DESCRIPTION
**Pull request by Kuri-su**
UCP #503

### What problem does this PR solve? 

fix #503 

### What is changed and how it works?

* add pkg 'metricsProxy' 
  * it will note labels list when operate metrics 
* add some code to clear metrics labels value in 'Close' function in these Struct
  * SubTask
  * Dumpling
  * Loader
  * Syncer
  * Mydumper

### CODE REVIEW

**Comment by Kuri-su** [comment]
Maybe i need to add unit tests in this pkg

**Comment by Kuri-su** [comment]
@csuzhangxc PTAL :smile:

**Comment by Kuri-su** [comment]
@csuzhangxc I'am Done , PTAL :smile:

File: pkg/metricsproxy/histogramVec_test.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	for _, oneCase := range testCases {
+		histogram := NewHistogramVec(prometheus.HistogramOpts{
+			Namespace:   "dm",
+			Subsystem:   "metricsProxy",
+			Name:        "Test_Histogram",
+			Help:        "dm histogram metrics proxy test",
+			ConstLabels: nil,
+		}, oneCase.LabelsNames)
+		for _, aArgs := range oneCase.AddArgs {
+			if rand.Intn(199)%2 == 0 {
```
Comment (lichunzhu) [inline pkg/metricsproxy/histogramVec_test.go:21]: Why `rand` is used here?

Comment (Kuri-su) [reply pkg/metricsproxy/histogramVec_test.go:21]: Just to improve test coverage, so I used a rand here, or is there any other recommended way

Comment (lichunzhu) [reply pkg/metricsproxy/histogramVec_test.go:21]: If you want to improve test coverage, `rand.Seed` should be called before use `rand.Intn`. Or the value of `rand.Intn` won't be updated.

Comment (Kuri-su) [reply pkg/metricsproxy/histogramVec_test.go:21]: or maybe i need to test two case in differemt method?

Comment (Kuri-su) [reply pkg/metricsproxy/histogramVec_test.go:21]: :joy: ok, got it

Comment (lichunzhu) [reply pkg/metricsproxy/histogramVec_test.go:21]: I think that's also good.

Comment (Kuri-su) [reply pkg/metricsproxy/histogramVec_test.go:23]: i fixed it

File: syncer/syncer.go [hunk 2]
```diff
@@ -2205,6 +2205,8 @@ func (s *Syncer) Close() {
 	s.Lock()
 	defer s.Unlock()
 
+	s.removeLabelValuesWithTaskInMetrics(s.cfg.Name)
```
Comment (csuzhangxc) [inline syncer/syncer.go:2208]: How about putting it in a later position (before other parts already stopped, like `s.closed.Set(true)`)? if so, ditto for other `remove***` methods.

File: pkg/metricsproxy/counterVec.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
+package metricsproxy
```
[the remark this answers is not in the document]
Comment (Kuri-su) [reply pkg/metricsproxy/counterVec.go]: ok, it's done. 

and is it a norm to use Camel-Case for file names?

maybe i need to learn the golang coding norm from Uber?

File: pkg/metricsproxy/proxy.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Proxy Interface
+type Proxy interface {
+	GetLabels() map[string]map[string]string
+	vecDelete(prometheus.Labels) bool
+}
+
+// noteLabels common function in Proxy
```
[the remark this answers is not in the document]
Comment (Kuri-su) [reply pkg/metricsproxy/proxy.go]: ok, it's done. maybe ;-)

File: syncer/syncer.go (hunk 2 above)
Comment (Kuri-su) [reply syncer/syncer.go]: done

File: pkg/metricsproxy/counterVec.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/metricsproxy/counterVec.go]: just we don't use Camel-Case for file names/

**Pull request merged**